build: add cmake install commands

ArcEarth opened the pull request:
The project is currently spending 5 minutes every time I press the build button, after linking CommunityShaders.dll, without any prompt. 

Turns out lot of this was the cost from all the `POST_BUILD`` commands attached to CommunityShaders, with `ZIP_TO_DIST` default to `ON`. 

This change splitted the expensive zip packaging command into seperate Package-XXX targets, where developers can manually build before release.

Beside, this change also added proper `cmake --install` support. You can use it to install the AIO package into any folder with `cmake --install ./build/ALL --prefix <TARGET>` command.

Finally, `COMMENTS` have been added to all commands, prompting what is running, building, copying or zipping.

With all these change, the default preset can now finish an incremental build in about **10 seconds** instead of 5 minutes.

The Package targets in Visual Studio CMake Targets View:
<img width="881" height="1434" alt="image" src="https://github.com/user-attachments/assets/5823380c-71bf-4f7f-a8d8-9fcb3848ecb9" />

Incremental build now takes ~10 seconds to complete (with AIO folder update), and show prompts:
<img width="3521" height="563" alt="image" src="https://github.com/user-attachments/assets/845fa377-4d6c-43bc-bf1a-5d129864ec7e" />


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - New packaging targets for AIO, Core, and per-feature packages producing stable, stamp-driven archives and supporting incremental AIO builds.
  - Standalone shader preparation and copy-shaders targets for CI/local workflows and a preparatory AIO target.
  - Improved automated/manual AIO packaging and clearer build-time messaging.

- Chores
  - Build presets reworked (Dev, Package, Shaders, Debug, unified ALL) and default install prefix adjusted to local AIO path.
  - Added environment-guard guidance for auto-deployment.

- Documentation
  - Expanded README and developer guides: build, packaging, Docker, debugging, and packaging usage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ArcEarth committed:
optimize(build): build nolonger takes 5 minutes to complete
The project is currently spending 5 minutes every time I press the build
button, after linking CommunityShaders.dll, without any prompt.
Turns out lot of this was the cost from all the `POST_BUILD`` commands attached to
CommunityShaders, with `ZIP_TO_DIST` default to `ON`.
This change splitted the expensive zip packaging command into seperate
Package-XXX targets, where developers can manually build
before release.
Beside, this change also added proper `cmake --install` support.
You can use it to install the AIO package into any folder with `cmake
--install ./build/ALL --prefix <TARGET>` command.

With all these change, the default preset can now finish an incremental build in about
10 seconds instead of 5 minutes.

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

ArcEarth committed:
Fix AI review
- Set `AIO_DIR`` earlier before `prepare_shaders``
- Ensure `dist` folder exists

ArcEarth commented on CMakeLists.txt:
```diff
@@ -191,122 +198,118 @@
+	COMMAND ${CMAKE_COMMAND} --install ${CMAKE_BINARY_DIR} --prefix ${FEATURE_PATH} --component SKSE
+	COMMAND ${CMAKE_COMMAND} -E copy_directory ${CORE_FEATURE_PATHS} ${FEATURE_PATH}
+	COMMAND ${CMAKE_COMMAND} -E rm -f -- ${FEATURE_PATH}/Core
+	COMMAND ${CMAKE_COMMAND} -E tar cfv ${CORE_PACKAGE} --format=7zip -- .
```
Thats a false claim, copy directory support multiple src.

ArcEarth commented on CMakeLists.txt:
```diff
@@ -191,122 +198,118 @@
+set(CORE_PACKAGE "${DIST_PATH}/${PROJECT_NAME}-${UTC_NOW}.7z")
 
-		list(APPEND DEPLOY_TARGET_HASHES ${DEPLOY_TARGET_HASH}.stamp)
+file(GLOB_RECURSE CORE_SOURCES CONFIGURE_DEPENDS "${CMAKE_SOURCE_DIR}/package/*")
```
valid, making change

ArcEarth commented on CMakeLists.txt:
```diff
@@ -191,122 +198,118 @@
+	COMMAND ${CMAKE_COMMAND} --install ${CMAKE_BINARY_DIR} --prefix ${AIO_DIR} --component Shaders
+	COMMAND ${CMAKE_COMMAND} -E touch copy_shaders.stamp
+	DEPENDS ${HLSL_FILES}
+	COMMENT "Copy shaders"
```
We already have the output stamp file. No need for byproducts here.

ArcEarth commented on CMakeLists.txt:
```diff
@@ -1,11 +1,18 @@
 project(
 	CommunityShaders
 	VERSION 1.3.4
 	LANGUAGES CXX
 )
 
+# default install path
+if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+  set_property(CACHE CMAKE_INSTALL_PREFIX PROPERTY VALUE "${CMAKE_CURRENT_BINARY_DIR}/aio")
+endif()
+
```
valid, making changes.

alandtse commented:
Please also read the spec for conventional commits. Your title is invalid.

alandtse reviewed (changes requested):
This breaks the ci as there is expectation that when the presets are set files are copied to certain locations for packaging. Please make the necessary changes to address those scripts and tools or do not alter core behavior.

ArcEarth commented:
Thanks, will take some time to test and fix CI later today :)

ArcEarth committed:
Fixing CI
Updating README.md for build instructions
Updating CMakePresets.json, removing unused presets like SE/AE
Change AIO folder from POST_BUILD to custom target

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

ArcEarth committed:
Revert renaming of CMakePresets because CI cannot be updated

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

alandtse commented:
> Hi, @alandtse, wondering how can I test changes to the CI script? I changed the CI script to use the `Package` build preset but it seems still using the old `ALL` preset >_<
> I have added the `AIO-Package` target back to `ALL` for now for CI, and added the lean preset as `Dev` to see if I can pass CI :)

It's building using the parameters for the cmake GitHub action. Run-cmake I believe. To test ci you may have to use a fork and not the main repo.

ArcEarth committed:
Revert CI files and testing shaders

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

alandtse reviewed (changes requested):
1. Please make sure you review the bot suggestions and resolve them after you've thought about it. The bot could be wrong, but sometimes it's actually right.
2. Note, not all of us us MO2.
3. Some of us deploy to multiple versions of skyrim so need the deployment scripts to work into multiple directories without further copying. I didn't double check copy_shaders, but that is used by some of us to automatically rebuild things in game. If it's not able to copy into a live skyrim folder, it's not working.
4. All commands should work from command line including to generate features and the aio we distribute for testing. I never log into Visual Studio so will not see the install targets.  The CI always builds via CLI and needs to generate all feature zips and aio zip.

alandtse committed:
build: merge upstream/dev into PR #1372

Brings PR #1372 up to date with upstream/dev (v1.4.6), integrating install
target improvements with upstream's deployment system.

Changes from PR #1372:
- Add CMake install() infrastructure for manual deployments
- Add Package-Core, Package-AIO-Manual, and Package-<Feature> targets
- Support cmake --install with custom --prefix
- Set CMAKE_INSTALL_PREFIX default to build/aio
- Add CMake policy CMP0116 settings

Changes from upstream/dev (v1.4.6):
- Add directx-headers, XeSS-SDK, Detours dependencies
- Implement incremental copy system with copy_if_different
- Add Windows robocopy integration for efficient deployment
- Add PREPARE_AIO target with intelligent dependency tracking
- Improve shader copying to preserve timestamps
- Multiple feature improvements and bug fixes

Conflict resolution:
- Unified AIO_DIR definition
- Kept ZIP_TO_DIST and AIO_ZIP_TO_DIST options
- Added install() commands alongside existing POST_BUILD system
- Maintained prepare_shaders target for CI
- AUTO_PLUGIN_DEPLOYMENT uses upstream robocopy approach

Tested:
- cmake --preset ALL configuration
- cmake --build build/ALL
- prepare_shaders target (CI critical)
- Package-Core target
- cmake --install with custom --prefix

alandtse committed:
build: make AIO archive incremental (stamp file)

Create a stable stamp file for the automatic AIO archive so CMake
only rebuilds the archive when inputs change. The AIO tar step now
writes ${CMAKE_CURRENT_BINARY_DIR}/aio_package.stamp and
`AIO_ZIP_PACKAGE` depends on that stamp while preserving the UTC
timestamp in the archive filename.

alandtse committed:
chore: address ai comments

- Replace "\"${FEATURE_PATH}/\"" with "${FEATURE_PATH}/" so feature paths are normal strings during packaging.
- Update README: prefer Developer PowerShell/x64 Native Tools prompt; use presets where appropriate and fix typos.

alandtse committed:
build: use target-file generator expression

- Replace literal project name in CORE_SOURCES with "$<TARGET_FILE:${PROJECT_NAME}>" so add_custom_command DEPENDS references the actual DLL path at build time.
